### DISCUSSION TITLE
Melos for private packages and multiple apps?

### DISCUSSION THREAD

**Discussion by scopendo** [invertase/melos#299] [Q&A]
Hi – I'm unclear whether melos is intended for monorepos containing multiple apps, as opposed to packages that might end up being published on pub.dev.

We currently have a monorepo that has multiple private packages in it, along with a couple of Flutter apps.  Some of the private packages depend on the other private packages, and all of the private packages are used either directly or indirectly by the Flutter apps, which are of course technically packages themselves.

The problem we're having with managing these in a single repository is knowing when to build the apps on the CI server – if we change app A in the repo, we want to build app B.  Sometimes, when we change a common package dependency, we still only want to build app A as that's what's currently being worked on with an upcoming release.

This is getting messy and so we were planning on splitting the apps into separate repositories and having a single repo for our shared private packages.  Unfortunately, Dart doesn't seem to support apps referencing the private packages via a git URL when those private packages reference each other via a relative path (see https://github.com/dart-lang/pub/issues/2447).

Does melos help here, and if so, does it require all the packages to be in a monorepo or can some of the packages be in separate git repositories?

Thanks,
Scott

**Comment by blaugold** [comment]
> Hi – I'm unclear whether melos is intended for monorepos containing multiple apps, as opposed to packages that might end up being published on pub.dev.

Melos should be able to support both use cases, but has probably been used mostly for the later one. There might functionality required for app development that is missing, so please let us know when that is the case.

 > The problem we're having with managing these in a single repository is knowing when to build the apps on the CI server – if we change app A in the repo, we want to build app B.

I guess you meant "we don't want to build app B"? Running tasks only for packages affected by a change is something we are working on.

> Sometimes, when we change a common package dependency, we still only want to build app A as that's what's currently being worked on with an upcoming release.

I'm not certain on the details, but that's probably something that will be possible.

> This is getting messy and so we were planning on splitting the apps into separate repositories and having a single repo for our shared private packages. Unfortunately, Dart doesn't seem to support apps referencing the private packages via a git URL when those private packages reference each other via a relative path (see https://github.com/dart-lang/pub/issues/2447).

Melos supports [git dependencies between packages](https://github.com/invertase/melos/issues/159) in the same repo and can update the git tags accordingly when running `melos version`.
